util_for_test.go

```diff
@@ -0,0 +1,304 @@
+				t.Fatalf("data mismatch:\n    want: %+v, got:  %+v", want, d)
+			}
+		})
+	}
+}
+
+type DBForTest struct {
+	*DB
+}
+
+func NewTestSurrealDB(t testing.TB) (string, *DBForTest, func()) {
+	t.Helper()
+	checkSurrealCLIOnce.Do(func() {
+		if _, err := exec.LookPath("surreal"); err == nil {
+			testHasSurrealCLI = true
+		}
+	})
+	if !testHasSurrealCLI {
+		t.Skip("surreal CLI not found, skipping")
+	}
+
+	port := getFreePort(t)
+	srvEndpoint := fmt.Sprintf("0.0.0.0:%d", port)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, "surreal",
+		"start",
+		"--auth",
+		"--user", "root",
+		"--pass", "surrealdb",
+		"--bind", srvEndpoint,
+		"memory")
+	// Ensure to wait for a short while before context cancellation to
+	// propagate, so that the client can shut down before.
+	// NOTE: There is currently a race condition and potential panic due to the
+	// closed websocket being used in a tight loop.
+	cmd.WaitDelay = DelayBeforeServerExit
+
+	go func() {
+		err := cmd.Run()
+		if err != nil {
+			if errors.Is(ctx.Err(), context.Canceled) {
+				t.Logf("SurrealDB server shut down by context")
+				return
+			}
+			exit := &exec.ExitError{}
+			if errors.As(err, &exit) {
+				t.Errorf("SurrealDB server exited with error: %v", err)
+				return
+			}
+			t.Errorf("Command failed to run: %v", err)
+		}
+	}()
+
+	clientEndpoint := fmt.Sprintf("localhost:%d", port)
+
+	// Give some time for the server to start up.
+	time.Sleep(DelayAfterServerStart)
+	db := newDBForTest(t, clientEndpoint)
+	login := &Auth{
+		Username:  "root",
+		Password:  "surrealdb",
+		Namespace: "dummy",
+		Database:  "dummy",
+	}
+	if _, err := db.Signin(login); err != nil {
+		t.Fatalf("failed to connect using '%+v' for signin: %v", login, err)
+	}
+
+	// Client can potentially stay around, so ensure to close connection after
+	// the test is complete.
+	t.Cleanup(func() {
+		db.conn.Close()
+	})
+
+	return clientEndpoint, &DBForTest{db}, cancel
+}
+
+func getFreePort(t testing.TB) uint32 {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("could not find any open port: %v", err)
+	}
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		t.Fatalf("could not listen to \"%s\": %v", addr, err)
+	}
+	defer l.Close()
+	return uint32(l.Addr().(*net.TCPAddr).Port)
+}
+
+func newDBForTest(t testing.TB, endpoint string) *DB {
+	url := fmt.Sprintf("ws://%s/rpc", endpoint)
+	buff := bytes.NewBuffer([]byte{})
+	handler := rawslog.NewJSONHandler(buff, &rawslog.HandlerOptions{Level: rawslog.LevelDebug})
+	log := slog.New(handler)
+	ws, err := gorilla.Create().SetTimeOut(5 * time.Second).SetCompression(true).Logger(log).Connect(url)
+	if err != nil {
+		t.Fatalf("failed to create websocket connection: %v", err)
+	}
+	db, err := New(url, ws)
+	if err != nil {
+		t.Errorf("failed to establish connection to \"%v\": %v", url, err)
+	}
+	return db
+}
+
+func (d *DBForTest) Prepare(t testing.TB, schema string) {
```

Comment:
It is a good point, and I'm absolutely fine dropping this. If we are to drop this, we would want to add simple test cases where we can showcase how the database prep can be done.

Just to give a bit more context and clarity, this code actually comes from my own testing with https://github.com/upsidr/surrealtest, which uses [Ory's dockertest](https://github.com/ory/dockertest) package to start a new database. I wanted to set up the tables before going into the main code to be tested, which can fail with `t.Fatal`, so that I don't get confused about when test error is due to some prep SurrealQL syntax error, rather than code issues.
> Side Note: Although not the same syntax, Google Cloud Spanner has `spannertest` with a dedicated function `UpdateDDL` https://pkg.go.dev/cloud.google.com/go/spanner/spannertest#Server.UpdateDDL to set up the database. This is because `spannertest` is only an emulator and not a real database, but this makes it clear that any error from `UpdateDDL` would be a fatal one and does not have to do with the actual test case.

This is certainly some additional code to be maintained, but test helpers like this take in `t.testing.TB`, so the usage would be limited. Having testify and using suite is about the same IMHO -- they add some complexity / extra layer for convenience. I believe this is more of a preference, and I'm absolutely happy to go with whichever route.